src/core/dep-graph.ts

```diff
@@ -327,26 +303,41 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
+  ): types.PkgInfo[][] {
     const parentNodesIds = this.getNodeParentsNodeIds(nodeId);
+    const pkgInfo = this.getNodePkg(nodeId);
+
+    if (ancestors.includes(nodeId)) {
+      return [];
+    }
+
     if (parentNodesIds.length === 0) {
-      return [[this.getNodePkg(nodeId)]];
+      return [[pkgInfo]];
     }
 
     const allPaths: types.PkgInfo[][] = [];
-    parentNodesIds.map((id) => {
-      const out = this.pathsFromNodeToRoot(id).map((path) => {
-        return [this.getNodePkg(nodeId)].concat(path);
-      });
-      for (const path of out) {
-        allPaths.push(path);
-      }
-    });
+    for (const id of parentNodesIds) {
+      allPaths.push(
+        ...this.pathsFromNodeToRoot(id, ancestors.concat(nodeId)).map((path) =>
+          [pkgInfo].concat(path),
+        ),
+      );
+    }
 
     return allPaths;
   }
 
-  private countNodePathsToRoot(nodeId: string): number {
+  private countNodePathsToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
+  ): number {
+    if (ancestors.includes(nodeId)) {
```

Comment:
Again, this is weird. Surely you want this before recursing into `countNodePathsToRoot` not afterwards